src/chords.py

```diff
@@ -0,0 +1,164 @@
+import queue
+
+import numpy as np
+import pandas as pd
+import fortepyan as ff
+from tqdm import tqdm
+import chorder.dechorder
+import chorder.timepoints
+from datasets import load_dataset
+from miditoolkit.midi.containers import Note
+from pretty_midi.pretty_midi import PrettyMIDI
+
+from helpers import plot
+
+
+def start_to_ticks(row, piece: PrettyMIDI):
+    return piece.time_to_tick(row["start"])
+
+
+def end_to_ticks(row, piece: PrettyMIDI):
+    return piece.time_to_tick(row["end"])
+
+
+def end_to_time(row, piece: PrettyMIDI):
+    return piece.tick_to_time(row["end"])
+
+
+def get_chords(piece: ff.MidiPiece):
+    """
+    Extracts chords from a MidiPiece object.
+    Treats every group of simultaneously playing notes,
+    which can be interpreted as a chord by a dechorder, as a chord.
+
+    Parameters:
+        piece (MidiPiece): A MidiPiece object containing musical events.
+
+    Returns:
+        list: A list of detected chords in the format [chord_name, start_time, end_time].
+    """
+    # mido_obj = midipiece_to_midifile(piece)
+    # define comparator
+    setattr(Note, "__lt__", lambda self, other: self.end <= other.end)
+    pq = queue.PriorityQueue()
+    start = 0
+    count = 0
+    chords = []
+    prev_chord = None
+    midi_data = piece.df
+    piece = piece.to_midi()
+    midi_data["start"] = midi_data.apply(start_to_ticks, axis=1, args=(piece,))
+    midi_data["end"] = midi_data.apply(end_to_ticks, axis=1, args=(piece,))
+    notes_midi = np.array(midi_data[["velocity", "pitch", "start", "end"]])
+
+    for note in notes_midi:
+        note = Note(*note)
+        pq.put(note)
+        end = note.end
+        notes = [note for note in pq.queue]
+        # Pop notes which ended before current note started
+        if pq.empty():
+            continue
+        while note.start > pq.queue[0].end:
```

Comment:
I'm not sure if I'm reading this correctly, but it looks like as long as the first note of the sequence is being hold, everything after it is going to be considered part of the chord.

For example, this gets recognized as CM:

```
   pitch  start  end  duration  velocity
0     60      0    3         3        80
1     64      1    3         2        80
2     67      2    3         1        80
```

But there is one second distance between each of the notes, so the expected output would be "no chords".

The original intention was to find chords defined as notes performed simultaneously - the issue here being that in reality those types of chords are spread over time - either due to aesthetic choices of the player, or just because it's almost impossible to press multiple keys at the exactly same time with millisecond precision :)

Your logic can be adjusted for those requirements with some kind of thresholding for the maximum allowed duration of the chord:

```python
while pq.queue[0].end - note.start  < threshold:
```